File: 0221.MaximalSquare.go
func maximalSquare(matrix [][]byte) int {
    if len(matrix) == 0 {
        return 0
    }
    m, n := len(matrix)+1, len(matrix[0])+1
    dp := make([][]int, m)
    for i := range dp {
        dp[i] = make([]int, n)
    }
    Len := 0
    for i := 1; i < m; i++ {
        for j := 1; j < n; j++ {
            if matrix[i-1][j-1] == '1' {
                dp[i][j] = min(dp[i-1][j-1], min(dp[i-1][j], dp[i][j-1])) + 1
                if dp[i][j] > Len {
                    Len = dp[i][j]
                }
            }
        }
    }
    return Len*Len
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

File: 0000-0999/0221.MaximalSquare.py
class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        if not matrix:
            return 0
        row = len(matrix) + 1
        col = len(matrix[0]) + 1
        dp = [[0]*col for i in xrange(row)]
        maxLen = 0
        for i in xrange(1, row):
            for j in xrange(1, col):
                if matrix[i-1][j-1] == '1':
                    dp[i][j] = min(dp[i-1][j-1], dp[i-1][j], dp[i][j-1]) + 1
                    maxLen = max(maxLen, dp[i][j])
        return maxLen**2
